### PR TITLE
reconstruct memory when set EnableMemory to false in tracing

### PR DESCRIPTION
**Pull request by lightsing**
### OpCodes Impl
- [x] CALLDATACOPY
- [x] RETURNDATACOPY
- [x] CODECOPY
- [x] EXTCODECOPY
- [x] MSTORE
- [x] MSTORE8
- [x] RETURN

### Unit Tests
- [x] CALLDATACOPY
- [x] RETURNDATACOPY
- [x] CODECOPY
- [x] EXTCODECOPY
- [x] MSTORE
- [x] MSTORE8
- [x] RETURN

### Integration Test
- [x] Test with tracer config: `EnableMemory: false`

### Note
this pr is related to scroll-tech/go-ethereum/issues/94 and privacy-scaling-explorations/zkevm-circuits/issues/502.

### CODE REVIEW

File: bus-mapping/src/evm/opcodes.rs [hunk 1]
```diff
@@ -73,155 +81,164 @@ pub trait Opcode: Debug {
...
+        OpcodeId::REVERT => &Return,
         OpcodeId::SELFDESTRUCT => {
             warn!("Using dummy gen_selfdestruct_ops for opcode SELFDESTRUCT");
-            dummy_gen_selfdestruct_ops
+            &DummySelfDestruct
         }
         OpcodeId::CALLCODE | OpcodeId::DELEGATECALL | OpcodeId::STATICCALL => {
             warn!("Using dummy gen_call_ops for opcode {:?}", opcode_id);
-            dummy_gen_call_ops
+            &DummySelfDestruct
```
Comment (icemelon) [inline bus-mapping/src/evm/opcodes.rs:233]: why `DummySelfDestruct` for these call ops?

File: bus-mapping/src/evm/opcodes.rs [hunk 2]
```diff
@@ -534,149 +570,167 @@ pub fn gen_end_tx_ops(
...
 }
 
-fn dummy_gen_call_ops(
-    state: &mut CircuitInputStateRef,
-    geth_steps: &[GethExecStep],
-) -> Result<Vec<ExecStep>, Error> {
-    let geth_step = &geth_steps[0];
-    let mut exec_step = state.new_step(geth_step)?;
-
-    let tx_id = state.tx_ctx.id();
```
Comment (lispc) [inline bus-mapping/src/evm/opcodes.rs:544]: looks like git failed to find the unchanged part...

Comment (lispc) [reply bus-mapping/src/evm/opcodes.rs:544]: oh no, we made one more level of indent

File: bus-mapping/src/evm/opcodes.rs [hunk 3]
```diff
@@ -233,8 +250,27 @@ pub fn gen_associated_ops(
...
 ) -> Result<Vec<ExecStep>, Error> {
-    let fn_gen_associated_ops = fn_gen_associated_ops(opcode_id);
-    fn_gen_associated_ops(state, geth_steps)
+    let opcode = down_cast_to_opcode(opcode_id);
+    if geth_steps.len() > 1 && opcode_id.need_reconstruction() {
+        let memory = opcode.reconstruct_memory(state, geth_steps)?;
+        if geth_steps[1].memory.borrow().is_empty() {
+            geth_steps[1].memory.replace(memory.clone());
+        } else {
+            assert_eq!(&memory, geth_steps[1].memory.borrow().deref());
```
Comment (lispc) [inline bus-mapping/src/evm/opcodes.rs:259]: add comments here:

line 257 happens when memory trace disabled, while line 259  enabled.
line 259 is for debugging for next several months. In long term it is not needed

**Comment by lispc** [comment]
very clean codes.

After addressing the above comments, i think it can be sent to upstream for the first round of reviewing

**Comment by lispc** [comment]
i will merge this branch into https://github.com/scroll-tech/zkevm-circuits/pull/166 soon and use it to fuzz test a lot of mainnet blocks

**Comment by lispc** [comment]
hmm.. i clone the branch locally and i found i have to change some `memory: Memory::from(...)` to `memory: RefCell::new(Memory::from...)` in eth-types/src/lib.rs to pass compiling... strange that the CI passed.

**Comment by lispc** [comment]
is the name "reconstruct_memory" good enough? or change it to: 

     -> reconstruct_memory_for_next_step 
or 
     -> replay_memory_updates
?
@icemelon

File: bus-mapping/src/evm/opcodes.rs (hunk 1 above)
Comment (lightsing) [reply bus-mapping/src/evm/opcodes.rs:233]: it's a typo

**Comment by lightsing** [comment]
I add some debug output. Here's the most common failed reason `caller id not found in call map`.

```
test evm_circuit::execution::begin_tx::test::begin_tx_gadget_rand ... FAILED
test evm_circuit::execution::begin_tx::test::begin_tx_gadget_simple ... FAILED
test evm_circuit::execution::call::test::call_gadget_nested ... FAILED
test evm_circuit::execution::call::test::call_gadget_simple ... FAILED
test evm_circuit::execution::calldatacopy::test::calldatacopy_gadget_multi_step ... FAILED
test evm_circuit::execution::calldatacopy::test::calldatacopy_gadget_out_of_bound ... FAILED
test evm_circuit::execution::calldataload::test::calldataload_gadget_internal ... FAILED
test evm_circuit::execution::calldatacopy::test::calldatacopy_gadget_simple ... FAILED
test evm_circuit::execution::calldatacopy::test::calldatacopy_gadget_zero_length ... FAILED
test evm_circuit::execution::sload::test::sload_gadget_rand ... FAILED
test evm_circuit::execution::sload::test::sload_gadget_simple ... FAILED
test evm_circuit::execution::sstore::test::sstore_gadget_delete_slot ... FAILED
test evm_circuit::execution::sstore::test::sstore_gadget_no_refund ... FAILED
test evm_circuit::execution::sstore::test::sstore_gadget_recreate_slot ... FAILED
test evm_circuit::execution::sstore::test::sstore_gadget_recreate_slot_and_reset_inexistent ... FAILED
test evm_circuit::execution::sstore::test::sstore_gadget_reset_existing ... FAILED
test evm_circuit::execution::sstore::test::sstore_gadget_reset_inexistent ... FAILED

failures:

---- evm_circuit::execution::begin_tx::test::begin_tx_gadget_rand stdout ----
[zkevm-circuits/src/evm_circuit/execution/begin_tx.rs:376] random_amount = 219011809505302761
[zkevm-circuits/src/evm_circuit/execution/begin_tx.rs:376] random_gas_price = 1443137446
[bus-mapping/src/evm/opcodes/return.rs:24] &state.block_ctx.call_map = {
    1: (
        0,
        0,
    ),
}
[bus-mapping/src/evm/opcodes/return.rs:27] &current_call.caller_id = 0
thread 'evm_circuit::execution::begin_tx::test::begin_tx_gadget_rand' panicked at 'caller id not found in call map', bus-mapping/src/evm/opcodes/return.rs:32:18
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- evm_circuit::execution::begin_tx::test::begin_tx_gadget_simple stdout ----
[bus-mapping/src/evm/opcodes/return.rs:24] &state.block_ctx.call_map = {
    1: (
        0,
        0,
    ),
}
[bus-mapping/src/evm/opcodes/return.rs:27] &current_call.caller_id = 0
thread 'evm_circuit::execution::begin_tx::test::begin_tx_gadget_simple' panicked at 'caller id not found in call map', bus-mapping/src/evm/opcodes/return.rs:32:18

---- evm_circuit::execution::call::test::call_gadget_nested stdout ----
[bus-mapping/src/circuit_input_builder/input_state_ref.rs:545] &call = Call {
    call_id: 30,
    caller_id: 1,
    kind: Call,
    is_static: false,
    is_root: false,
    is_persistent: true,
    is_success: true,
    rw_counter_end_of_reversion: 0,
    caller_address: 0xfefefefefefefefefefefefefefefefefefefefe,
    address: 0xffffffffffffffffffffffffffffffffffffffff,
    code_source: Address(
        0xffffffffffffffffffffffffffffffffffffffff,
    ),
    code_hash: 0xd003426e799329b8dca093f3bbab55a5e4e9f3c40160fc942068eef712ae88ad,
    depth: 2,
    value: 0,
    call_data_offset: 0,
    call_data_length: 0,
    return_data_offset: 0,
    return_data_length: 0,
}
[bus-mapping/src/evm/opcodes/return.rs:24] &state.block_ctx.call_map = {
    1: (
        0,
        0,
    ),
    30: (
        0,
        1,
    ),
}
[bus-mapping/src/evm/opcodes/return.rs:27] &current_call.caller_id = 1
[bus-mapping/src/circuit_input_builder/input_state_ref.rs:545] &call = Call {
    call_id: 83,
    caller_id: 1,
    kind: Call,
    is_static: false,
    is_root: false,
    is_persistent: true,
    is_success: true,
    rw_counter_end_of_reversion: 0,
    caller_address: 0xfefefefefefefefefefefefefefefefefefefefe,
    address: 0xffffffffffffffffffffffffffffffffffffffff,
    code_source: Address(
        0xffffffffffffffffffffffffffffffffffffffff,
    ),
    code_hash: 0xd003426e799329b8dca093f3bbab55a5e4e9f3c40160fc942068eef712ae88ad,
    depth: 2,
    value: 0,
    call_data_offset: 0,
    call_data_length: 0,
    return_data_offset: 0,
    return_data_length: 0,
}
[bus-mapping/src/evm/opcodes/return.rs:24] &state.block_ctx.call_map = {
    83: (
        0,
        2,
    ),
    1: (
        0,
        0,
    ),
    30: (
        0,
        1,
    ),
}
[bus-mapping/src/evm/opcodes/return.rs:27] &current_call.caller_id = 1
[bus-mapping/src/evm/opcodes/return.rs:24] &state.block_ctx.call_map = {
    83: (
        0,
        2,
    ),
    1: (
        0,
        0,
    ),
    30: (
        0,
        1,
    ),
}
[bus-mapping/src/evm/opcodes/return.rs:27] &current_call.caller_id = 0
thread 'evm_circuit::execution::call::test::call_gadget_nested' panicked at 'caller id not found in call map', bus-mapping/src/evm/opcodes/return.rs:32:18

---- evm_circuit::execution::call::test::call_gadget_simple stdout ----
[bus-mapping/src/circuit_input_builder/input_state_ref.rs:545] &call = Call {
    call_id: 30,
    caller_id: 1,
    kind: Call,
    is_static: false,
    is_root: false,
    is_persistent: true,
    is_success: true,
    rw_counter_end_of_reversion: 0,
    caller_address: 0xfefefefefefefefefefefefefefefefefefefefe,
    address: 0xffffffffffffffffffffffffffffffffffffffff,
    code_source: Address(
        0xffffffffffffffffffffffffffffffffffffffff,
    ),
    code_hash: 0xc5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470,
    depth: 2,
    value: 0,
    call_data_offset: 0,
    call_data_length: 0,
    return_data_offset: 0,
    return_data_length: 0,
}
[bus-mapping/src/circuit_input_builder/input_state_ref.rs:545] &call = Call {
    call_id: 61,
    caller_id: 1,
    kind: Call,
    is_static: false,
    is_root: false,
    is_persistent: true,
    is_success: true,
    rw_counter_end_of_reversion: 0,
    caller_address: 0xfefefefefefefefefefefefefefefefefefefefe,
    address: 0xffffffffffffffffffffffffffffffffffffffff,
    code_source: Address(
        0xffffffffffffffffffffffffffffffffffffffff,
    ),
    code_hash: 0xc5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470,
    depth: 2,
    value: 0,
    call_data_offset: 0,
    call_data_length: 0,
    return_data_offset: 0,
    return_data_length: 0,
}
[bus-mapping/src/evm/opcodes/return.rs:24] &state.block_ctx.call_map = {
    30: (
        0,
        1,
    ),
    61: (
        0,
        2,
    ),
    1: (
        0,
        0,
    ),
}
[bus-mapping/src/evm/opcodes/return.rs:27] &current_call.caller_id = 0
thread 'evm_circuit::execution::call::test::call_gadget_simple' panicked at 'caller id not found in call map', bus-mapping/src/evm/opcodes/return.rs:32:18

---- evm_circuit::execution::calldatacopy::test::calldatacopy_gadget_multi_step stdout ----
[bus-mapping/src/circuit_input_builder/input_state_ref.rs:545] &call = Call {
    call_id: 66,
    caller_id: 1,
    kind: Call,
    is_static: false,
    is_root: false,
    is_persistent: true,
    is_success: true,
    rw_counter_end_of_reversion: 0,
    caller_address: 0x000000000000000000000000000000000cafe111,
    address: 0x000000000000000000000000000000000cafe222,
    code_source: Address(
        0x000000000000000000000000000000000cafe222,
    ),
    code_hash: 0x45c14d30a01509fc14647d97006bc426571f2f9876d39f76a4d1f77e0f1fbd3e,
    depth: 2,
    value: 0,
    call_data_offset: 48,
    call_data_length: 112,
    return_data_offset: 0,
    return_data_length: 0,
}
thread 'evm_circuit::execution::calldatacopy::test::calldatacopy_gadget_multi_step' panicked at 'assertion failed: `(left == right)`
  left: `[00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, b7, 88, be, 89, af, 05, a0, c7, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00]`,
 right: `[00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00]`', bus-mapping/src/evm/opcodes.rs:260:17

---- evm_circuit::execution::calldatacopy::test::calldatacopy_gadget_out_of_bound stdout ----
[bus-mapping/src/circuit_input_builder/input_state_ref.rs:545] &call = Call {
    call_id: 66,
    caller_id: 1,
    kind: Call,
    is_static: false,
    is_root: false,
    is_persistent: true,
    is_success: true,
    rw_counter_end_of_reversion: 0,
    caller_address: 0x000000000000000000000000000000000cafe111,
    address: 0x000000000000000000000000000000000cafe222,
    code_source: Address(
        0x000000000000000000000000000000000cafe222,
    ),
    code_hash: 0x5eeb596a7c7c1e11813c44dd342c3ad900b5c234941356220928b397fc231ba8,
    depth: 2,
    value: 0,
    call_data_offset: 64,
    call_data_length: 32,
    return_data_offset: 0,
    return_data_length: 0,
}
thread 'evm_circuit::execution::calldatacopy::test::calldatacopy_gadget_out_of_bound' panicked at 'attempt to subtract with overflow', bus-mapping/src/evm/opcodes/calldatacopy.rs:51:37

---- evm_circuit::execution::calldataload::test::calldataload_gadget_internal stdout ----
[bus-mapping/src/circuit_input_builder/input_state_ref.rs:545] &call = Call {
    call_id: 66,
    caller_id: 1,
    kind: Call,
    is_static: false,
    is_root: false,
    is_persistent: true,
    is_success: true,
    rw_counter_end_of_reversion: 0,
    caller_address: 0x000000000000000000000000000000000cafe111,
    address: 0x000000000000000000000000000000000cafe222,
    code_source: Address(
        0x000000000000000000000000000000000cafe222,
    ),
    code_hash: 0xd3178cec3f148ffc5e90101c9fe44a5c40ccd7cdd304e78ef157d777a9c7f49a,
    depth: 2,
    value: 0,
    call_data_offset: 0,
    call_data_length: 32,
    return_data_offset: 0,
    return_data_length: 0,
}
thread 'evm_circuit::execution::calldataload::test::calldataload_gadget_internal' panicked at 'assertion failed: `(left == right)`
  left: `Err([ConstraintNotSatisfied { constraint: Constraint { gate: Gate { index: 48, name: "CALL" }, index: 248, name: "State transition constraint of memory_word_size" }, location: InRegion { region: Region { index: 6, name: "Execution step" }, offset: 96 }, cell_values: [(VirtualCell { name: "", column: Column { column_type: Advice, index: 23 }, rotation: 0 }, "1")] }])`,
 right: `Ok(())`', zkevm-circuits/src/evm_circuit/execution/calldataload.rs:310:9

---- evm_circuit::execution::calldatacopy::test::calldatacopy_gadget_simple stdout ----
[bus-mapping/src/circuit_input_builder/input_state_ref.rs:545] &call = Call {
    call_id: 66,
    caller_id: 1,
    kind: Call,
    is_static: false,
    is_root: false,
    is_persistent: true,
    is_success: true,
    rw_counter_end_of_reversion: 0,
    caller_address: 0x000000000000000000000000000000000cafe111,
    address: 0x000000000000000000000000000000000cafe222,
    code_source: Address(
        0x000000000000000000000000000000000cafe222,
    ),
    code_hash: 0x7c24c8c5cf353cbf276d400e57add2dcac99433339c05d294adc5ad036213b82,
    depth: 2,
    value: 0,
    call_data_offset: 64,
    call_data_length: 64,
    return_data_offset: 0,
    return_data_length: 0,
}
thread 'evm_circuit::execution::calldatacopy::test::calldatacopy_gadget_simple' panicked at 'assertion failed: `(left == right)`
  left: `[00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, d9, 3e, e7, be, f4, 66, 2e, 5a, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00]`,
 right: `[00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00]`', bus-mapping/src/evm/opcodes.rs:260:17

---- evm_circuit::execution::calldatacopy::test::calldatacopy_gadget_zero_length stdout ----
[bus-mapping/src/circuit_input_builder/input_state_ref.rs:545] &call = Call {
    call_id: 66,
    caller_id: 1,
    kind: Call,
    is_static: false,
    is_root: false,
    is_persistent: true,
    is_success: true,
    rw_counter_end_of_reversion: 0,
    caller_address: 0x000000000000000000000000000000000cafe111,
    address: 0x000000000000000000000000000000000cafe222,
    code_source: Address(
        0x000000000000000000000000000000000cafe222,
    ),
    code_hash: 0x1bda9e953a12a1093f34061ba13196539e7d44c4837d9107368a809e66d2d5a5,
    depth: 2,
    value: 0,
    call_data_offset: 64,
    call_data_length: 64,
    return_data_offset: 0,
    return_data_length: 0,
}
thread 'evm_circuit::execution::calldatacopy::test::calldatacopy_gadget_zero_length' panicked at 'assertion failed: `(left == right)`
  left: `Err([ConstraintNotSatisfied { constraint: Constraint { gate: Gate { index: 48, name: "CALL" }, index: 248, name: "State transition constraint of memory_word_size" }, location: InRegion { region: Region { index: 6, name: "Execution step" }, offset: 96 }, cell_values: [(VirtualCell { name: "", column: Column { column_type: Advice, index: 23 }, rotation: 0 }, "1")] }])`,
 right: `Ok(())`', zkevm-circuits/src/evm_circuit/execution/calldatacopy.rs:329:9

---- evm_circuit::execution::sload::test::sload_gadget_rand stdout ----
[bus-mapping/src/evm/opcodes/return.rs:24] &state.block_ctx.call_map = {
    1: (
        0,
        0,
    ),
}
[bus-mapping/src/evm/opcodes/return.rs:27] &current_call.caller_id = 0
thread 'evm_circuit::execution::sload::test::sload_gadget_rand' panicked at 'caller id not found in call map', bus-mapping/src/evm/opcodes/return.rs:32:18

---- evm_circuit::execution::sload::test::sload_gadget_simple stdout ----
[bus-mapping/src/evm/opcodes/return.rs:24] &state.block_ctx.call_map = {
    1: (
        0,
        0,
    ),
}
[bus-mapping/src/evm/opcodes/return.rs:27] &current_call.caller_id = 0
thread 'evm_circuit::execution::sload::test::sload_gadget_simple' panicked at 'caller id not found in call map', bus-mapping/src/evm/opcodes/return.rs:32:18

---- evm_circuit::execution::sstore::test::sstore_gadget_delete_slot stdout ----
[bus-mapping/src/evm/opcodes/return.rs:24] &state.block_ctx.call_map = {
    1: (
        0,
        0,
    ),
}
[bus-mapping/src/evm/opcodes/return.rs:27] &current_call.caller_id = 0
thread 'evm_circuit::execution::sstore::test::sstore_gadget_delete_slot' panicked at 'caller id not found in call map', bus-mapping/src/evm/opcodes/return.rs:32:18

---- evm_circuit::execution::sstore::test::sstore_gadget_no_refund stdout ----
[bus-mapping/src/evm/opcodes/return.rs:24] &state.block_ctx.call_map = {
    1: (
        0,
        0,
    ),
}
[bus-mapping/src/evm/opcodes/return.rs:27] &current_call.caller_id = 0
thread 'evm_circuit::execution::sstore::test::sstore_gadget_no_refund' panicked at 'caller id not found in call map', bus-mapping/src/evm/opcodes/return.rs:32:18

---- evm_circuit::execution::sstore::test::sstore_gadget_recreate_slot stdout ----
[bus-mapping/src/evm/opcodes/return.rs:24] &state.block_ctx.call_map = {
    1: (
        0,
        0,
    ),
}
[bus-mapping/src/evm/opcodes/return.rs:27] &current_call.caller_id = 0
thread 'evm_circuit::execution::sstore::test::sstore_gadget_recreate_slot' panicked at 'caller id not found in call map', bus-mapping/src/evm/opcodes/return.rs:32:18

---- evm_circuit::execution::sstore::test::sstore_gadget_recreate_slot_and_reset_inexistent stdout ----
[bus-mapping/src/evm/opcodes/return.rs:24] &state.block_ctx.call_map = {
    1: (
        0,
        0,
    ),
}
[bus-mapping/src/evm/opcodes/return.rs:27] &current_call.caller_id = 0
thread 'evm_circuit::execution::sstore::test::sstore_gadget_recreate_slot_and_reset_inexistent' panicked at 'caller id not found in call map', bus-mapping/src/evm/opcodes/return.rs:32:18

---- evm_circuit::execution::sstore::test::sstore_gadget_reset_existing stdout ----
[bus-mapping/src/evm/opcodes/return.rs:24] &state.block_ctx.call_map = {
    1: (
        0,
        0,
    ),
}
[bus-mapping/src/evm/opcodes/return.rs:27] &current_call.caller_id = 0
thread 'evm_circuit::execution::sstore::test::sstore_gadget_reset_existing' panicked at 'caller id not found in call map', bus-mapping/src/evm/opcodes/return.rs:32:18

---- evm_circuit::execution::sstore::test::sstore_gadget_reset_inexistent stdout ----
[bus-mapping/src/evm/opcodes/return.rs:24] &state.block_ctx.call_map = {
    1: (
        0,
        0,
    ),
}
[bus-mapping/src/evm/opcodes/return.rs:27] &current_call.caller_id = 0
thread 'evm_circuit::execution::sstore::test::sstore_gadget_reset_inexistent' panicked at 'caller id not found in call map', bus-mapping/src/evm/opcodes/return.rs:32:18


failures:
    evm_circuit::execution::begin_tx::test::begin_tx_gadget_rand
    evm_circuit::execution::begin_tx::test::begin_tx_gadget_simple
    evm_circuit::execution::call::test::call_gadget_nested
    evm_circuit::execution::call::test::call_gadget_simple
    evm_circuit::execution::calldatacopy::test::calldatacopy_gadget_multi_step
    evm_circuit::execution::calldatacopy::test::calldatacopy_gadget_out_of_bound
    evm_circuit::execution::calldatacopy::test::calldatacopy_gadget_simple
    evm_circuit::execution::calldatacopy::test::calldatacopy_gadget_zero_length
    evm_circuit::execution::calldataload::test::calldataload_gadget_internal
    evm_circuit::execution::sload::test::sload_gadget_rand
    evm_circuit::execution::sload::test::sload_gadget_simple
    evm_circuit::execution::sstore::test::sstore_gadget_delete_slot
    evm_circuit::execution::sstore::test::sstore_gadget_no_refund
    evm_circuit::execution::sstore::test::sstore_gadget_recreate_slot
    evm_circuit::execution::sstore::test::sstore_gadget_recreate_slot_and_reset_inexistent
    evm_circuit::execution::sstore::test::sstore_gadget_reset_existing
    evm_circuit::execution::sstore::test::sstore_gadget_reset_inexistent

test result: FAILED. 125 passed; 17 failed; 16 ignored; 0 measured; 0 filtered out; finished in 596.15s
```

**Comment by lispc** [comment]
i will debug them this afternoon..

**Comment by lispc** [comment]
add a LoggerConfig in TraceConfig in trace.go. So we can control EnableMemory in rust side.

**Comment by lispc** [comment]
the current CI is passed. And i saw there the mem trace had already been disabled?

File: eth-types/src/evm_types/memory.rs [hunk 4]
```diff
@@ -285,6 +319,18 @@ impl Memory {
     pub fn word_size(&self) -> usize {
         self.0.len() / 32
     }
+
+    /// Resize the memory for at least length and align to 32 bytes.
+    pub fn extend_at_least(&mut self, minimal_length: usize) {
```
Comment (icemelon) [inline eth-types/src/evm_types/memory.rs:324]:
```suggestion
    pub fn expand(&mut self, length: usize) {
```

File: eth-types/src/evm_types/opcode_ids.rs [hunk 5]
```diff
@@ -333,6 +333,21 @@ impl OpcodeId {
     pub fn is_log(&self) -> bool {
         self.as_u8() >= Self::LOG0.as_u8() && self.as_u8() <= Self::LOG4.as_u8()
     }
+    /// Returns `true` if the `OpcodeId` is a `RETURN` or `REVERT`.
```
Comment (icemelon) [inline eth-types/src/evm_types/opcode_ids.rs:336]: add empty line

File: eth-types/src/evm_types/opcode_ids.rs [hunk 6]
```diff
@@ -333,6 +333,21 @@ impl OpcodeId {
...
+    pub fn is_return(&self) -> bool {
+        matches!(self, OpcodeId::RETURN | OpcodeId::REVERT)
+    }
+    /// Returns `true` if the `OpcodeId` is a CALL-like.
```
Comment (icemelon) [inline eth-types/src/evm_types/opcode_ids.rs:340]: add empty line

File: eth-types/src/evm_types/opcode_ids.rs [hunk 7]
```diff
@@ -638,6 +653,20 @@ impl OpcodeId {
             OpcodeId::SELFDESTRUCT => GasCost::SELFDESTRUCT,
         }
     }
+
+    /// Returns if the `OpcodeId` needs reconstruction
+    pub const fn need_reconstruction(&self) -> bool {
```
Comment (icemelon) [inline eth-types/src/evm_types/opcode_ids.rs:658]:
```suggestion
    pub const fn has_memory_access(&self) -> bool {
```

File: eth-types/src/evm_types/memory.rs [hunk 8]
```diff
@@ -285,6 +319,18 @@ impl Memory {
...
+        if minimal_length > self.0.len() {
+            let resize = if minimal_length % 32 == 0 {
+                minimal_length
+            } else {
+                (minimal_length / 32 + 1) * 32
+            };
+            self.0.resize(resize, 0);
+        }
```
Comment (icemelon) [inline eth-types/src/evm_types/memory.rs:332]:
```suggestion
        let memory_size = (length + 31) / 32 * 32
        if memory_size > self.0.len() {
            self.0.resize(memory_size, 0);
        }
```

File: eth-types/src/evm_types/opcode_ids.rs [hunk 9]
```diff
@@ -333,6 +333,21 @@ impl OpcodeId {
     pub fn is_log(&self) -> bool {
         self.as_u8() >= Self::LOG0.as_u8() && self.as_u8() <= Self::LOG4.as_u8()
     }
+    /// Returns `true` if the `OpcodeId` is a `RETURN` or `REVERT`.
+    pub fn is_return(&self) -> bool {
```
Comment (icemelon) [inline eth-types/src/evm_types/opcode_ids.rs:337]: seems like this function is not used?

File: bus-mapping/src/evm/opcodes/return.rs [hunk 10]
```diff
@@ -1,24 +1,240 @@
...
+        } else {
+            Ok(geth_steps[0].memory.borrow().clone())
+        }
+    }
+}
 
-        state.handle_return(geth_step)?;
-        Ok(vec![exec_step])
+// TODO: circuit implement
+// fn gen_calldatacopy_step(
```
Comment (icemelon) [inline bus-mapping/src/evm/opcodes/return.rs:65]: remove these?

**Review by icemelon — COMMENTED** [review]
so this pr disables the memory tracing by default in all of the busmapping tests, right?

File: eth-types/src/evm_types/memory.rs (hunk 4 above)
Comment (lightsing) [reply eth-types/src/evm_types/memory.rs:324]: the old name is telling caller that this function might not expand the memory to given length. the length is aligned to 32 bytes.

**Comment by lispc** [comment]
@lightsing backup the branch, then squashed it into one commit. (I will merge this branch into other branch, so better clean the history)

File: bus-mapping/src/evm/opcodes/sha3.rs [hunk 11]
```diff
@@ -14,6 +14,8 @@ impl Opcode for Sha3 {
         state: &mut CircuitInputStateRef,
         geth_steps: &[GethExecStep],
     ) -> Result<Vec<ExecStep>, Error> {
+        // TODO: memory reads
```
Comment (lispc) [inline bus-mapping/src/evm/opcodes/sha3.rs:17]: "TODO: add memory read operations"

**Comment by lispc** [comment]
moved to https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/614

**Pull request closed without merging**